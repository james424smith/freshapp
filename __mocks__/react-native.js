import * as ReactNative from "react-native";

export * from "react-native";
export const Vibration = { vibrate: jest.fn() };
export const Adnimated = { spring: jest.fn() };

export default Object.setPrototypeOf(
  {
    Vibration,
    Adnimated,
  },
  ReactNative
);
