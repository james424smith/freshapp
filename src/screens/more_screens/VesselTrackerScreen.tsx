import React, { useState, useEffect, useCallback, useRef } from "react";
import { TouchableOpacity, View, Image, Platform } from "react-native";
import MapView, { PROVIDER_DEFAULT, Marker } from "react-native-maps";
import Text from "../../components/StyledText";
import { Icon } from "react-native-elements";
import mapPin from "../../../assets/images/ic_pin_small.png";
import styles from "../../styles/more_screens/VesselTrackerStyles";
import { useSelector } from "react-redux";
import { Assignment, VesselDetails } from "../../interfaces";
import { Region } from "../../interfaces/MapViewTypes";
import { IOS } from "../../constants/Platforms";

import { IRootReducerType } from "../../redux/reducers";
import valueOrDefault from "../../common/valueOrDefault";

const VesselTrackerScreen = () => {
  const map = useRef<MapView | null>(null);
  const [region, setRegion] = useState<Region>({
    latitude: 34.665199,
    longitude: 33.008535,
    latitudeDelta: 3,
    longitudeDelta: 3,
  });
  const [, updateState] = useState<number | undefined>();
  const forceUpdate = useCallback(() => updateState(1), []);
  const currentVesselDetails = useSelector<
    IRootReducerType,
    VesselDetails | undefined
  >(
    ({ assignmentDetails }) =>
      (valueOrDefault(
        assignmentDetails.assignmentDetails,
        []
      ) as Assignment[])[0]?.vesselDetails
  );
  const getVesselRegion = useCallback((): Region => {
    return {
      latitude: currentVesselDetails?.latitude ?? 34.665199,
      longitude: currentVesselDetails?.longitude ?? 33.008535,
      latitudeDelta: 3,
      longitudeDelta: 3,
    };
  }, [currentVesselDetails?.latitude, currentVesselDetails?.longitude]);

  useEffect(() => {
    setRegion(getVesselRegion());
  }, [getVesselRegion]);

  useEffect(() => {
    if (Platform.OS === IOS && map.current) {
      map.current.animateToRegion(getVesselRegion());
    }
  }, [getVesselRegion, map]);

  const animateToInitialRegion = (): void => {
    if (map.current) {
      map.current.animateToRegion(getVesselRegion());
    }
  };

  const onRegionChange = (regionLocal: Region): void => {
    setRegion(regionLocal);
  };

  return (
    <View style={styles.container}>
      <MapView
        style={styles.map}
        ref={map}
        provider={PROVIDER_DEFAULT}
        initialRegion={region}
        onRegionChange={(regionOnChange: Region) =>
          onRegionChange(regionOnChange)
        }
      >
        {currentVesselDetails?.latitude && currentVesselDetails?.longitude && (
          <Marker
            coordinate={{
              latitude: currentVesselDetails?.latitude,
              longitude: currentVesselDetails?.longitude,
            }}
            title={currentVesselDetails?.name}
          >
            <Image
              style={styles.img}
              source={mapPin}
              onLoad={() => forceUpdate()}
            />
          </Marker>
        )}
      </MapView>
      <View style={[styles.bubble, styles.latlng]}>
        <Text style={styles.textFormat}>
          {region.latitude.toPrecision(7)},{region.longitude.toPrecision(7)}
        </Text>
      </View>
      <View style={styles.buttonContainer}>
        <TouchableOpacity
          testID={"go-to-current-position"}
          onPress={() => animateToInitialRegion()}
          style={[styles.bubble, styles.button]}
        >
          <Icon name="my-location" />
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default VesselTrackerScreen;
