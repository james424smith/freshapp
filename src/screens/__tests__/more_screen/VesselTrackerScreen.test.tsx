import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react-native";
import VesselTrackerScreen from "../../more_screens/VesselTrackerScreen";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

const vesselDetails = {
  deadWeightTonnage: 7224,
  beam: 18.45,
  flag: "PMD",
  vesselImage: "1234",
  longitude: -0.243665,
  mainEngineOutput: 5300,
  managingAgent: "K & K",
  vesselSubType: "1ST GENERATION",
  grossTonnage: 6362,
  vesselType: "CONT",
  vesselCode: 863,
  lengthOverall: 121.76,
  draught: 6.69,
  vesselImoNumber: 9143972,
  mainEngineSeries: "L40/54",
  name: "WEC BRUEGHEL",
  mainEngineMaker: "MAN B&W",
  cargoInstallation: [
    [
      {
        label: "Number",
        key: "number",
        value: "2",
      },
      {
        label: "Unit",
        key: "unit",
        value: "TO",
      },
      {
        label: "Cargo Installation",
        key: "cargoInstallation",
        value: "CRANES",
      },
      {
        label: "Capacity",
        key: "capacity",
        value: "250",
      },
    ],
  ],
  latitude: 37.891758,
  lengthBetweenPerpendiculars: 114.9,
};

describe("VesselTrackerScreen snapshot test", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it("should render VesselTrackerScreen when vesselDetails is defined ", () => {
    const mockedState = {
      assignmentDetails: {
        assignmentDetails: [
          {
            vesselDetails,
          },
        ],
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON, getByTestId } = render(
      <Provider store={mockedStore(mockedState)}>
        <VesselTrackerScreen />
      </Provider>
    );

    fireEvent.press(getByTestId("go-to-current-position"));

    expect(toJSON()).toMatchSnapshot();
  });
  it("should render VesselTrackerScreen when vesselDetails is undefined", () => {
    const mockedState = {
      assignmentDetails: {
        assignmentDetails: [
          {
            vesselDetails: undefined,
          },
        ],
      },
      network: { isConnected: true },
    };
    const mockedStore = configureMockStore();

    const { toJSON } = render(
      <Provider store={mockedStore(mockedState)}>
        <VesselTrackerScreen />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
