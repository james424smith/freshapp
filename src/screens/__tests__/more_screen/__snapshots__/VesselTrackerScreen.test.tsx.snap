// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VesselTrackerScreen snapshot test should render VesselTrackerScreen when vesselDetails is defined  1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "bottom": 0,
      "justifyContent": "flex-end",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 20,
          "paddingHorizontal": 18,
          "paddingVertical": 12,
        },
        Object {
          "alignItems": "stretch",
          "width": 200,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#2A2E43",
            "fontFamily": "Roboto-Regular",
          },
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      37.89176
      ,
      -0.2436650
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "flexDirection": "row",
        "marginVertical": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 20,
          "justifyContent": "center",
          "marginHorizontal": 5,
          "opacity": 1,
          "paddingHorizontal": 8,
          "paddingVertical": 12,
          "width": 100,
        }
      }
      testID="go-to-current-position"
    >
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#242424",
                    "fontSize": 24,
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
              testID="iconIcon"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`VesselTrackerScreen snapshot test should render VesselTrackerScreen when vesselDetails is undefined 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "bottom": 0,
      "justifyContent": "flex-end",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 20,
          "paddingHorizontal": 18,
          "paddingVertical": 12,
        },
        Object {
          "alignItems": "stretch",
          "width": 200,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#2A2E43",
            "fontFamily": "Roboto-Regular",
          },
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      34.66520
      ,
      33.00854
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "flexDirection": "row",
        "marginVertical": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 20,
          "justifyContent": "center",
          "marginHorizontal": 5,
          "opacity": 1,
          "paddingHorizontal": 8,
          "paddingVertical": 12,
          "width": 100,
        }
      }
      testID="go-to-current-position"
    >
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#242424",
                    "fontSize": 24,
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
              testID="iconIcon"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
